builder/builder.go

```diff
@@ -0,0 +1,172 @@
+package builder
+
+import (
+	"encoding/json"
+	_ "os"
+
+	"github.com/ethereum/go-ethereum/common/hexutil"
+	"github.com/ethereum/go-ethereum/core/beacon"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/log"
+
+	"github.com/flashbots/go-boost-utils/bls"
+	boostTypes "github.com/flashbots/go-boost-utils/types"
+)
+
+type PubkeyHex string
+
+type ValidatorData struct {
+	Pubkey       PubkeyHex
+	FeeRecipient boostTypes.Address `json:"feeRecipient"`
+	GasLimit     uint64             `json:"gasLimit"`
+	Timestamp    uint64             `json:"timestamp"`
+}
+
+type IBeaconClient interface {
+	isValidator(pubkey PubkeyHex) bool
+	getProposerForNextSlot(requestedSlot uint64) (PubkeyHex, error)
+	onForkchoiceUpdate() (uint64, error)
+}
+
+type IRelay interface {
+	SubmitBlock(msg *boostTypes.BuilderSubmitBlockRequest) error
+	GetValidatorForSlot(nextSlot uint64) (ValidatorData, error)
+}
+
+type Builder struct {
+	beaconClient IBeaconClient
+	relay        IRelay
+
+	builderSecretKey     *bls.SecretKey
+	builderPublicKey     boostTypes.PublicKey
+	builderSigningDomain boostTypes.Domain
+}
+
+func NewBuilder(sk *bls.SecretKey, bc IBeaconClient, relay IRelay, builderSigningDomain boostTypes.Domain) *Builder {
+	pkBytes := bls.PublicKeyFromSecretKey(sk).Compress()
+	pk := boostTypes.PublicKey{}
+	pk.FromSlice(pkBytes)
+
+	_, err := bc.onForkchoiceUpdate()
+	if err != nil {
+		log.Error("could not initialize beacon client", "err", err)
```

Comment:
should this not abort here? (log.Fatal or return an error)

Reply:
Aborting here is problematic - I was doing it in the past, but sometimes you want geth to run anyway as you are bringing the BN up